### PR TITLE
[`DPOTrainer`] Fix peft + DPO + bf16 if one uses `generate_during_eval` or pre-computed logits

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do ?

Fixes https://github.com/huggingface/trl/issues/1202 which is a regression 

https://github.com/huggingface/trl/pull/1143 introduced `peft_module_casting_to_bf16` for `DPOTrainer` which worked fine when ` generate_during_eval` is set to `False`.

`peft_module_casting_to_bf16` will cast the LayerNorm layers in fp32 for stability purposes for smoother training, in case one uses peft + 4-bit quantization (QLoRA).

If one uses `bf16=True` in the `TrainingArguments`, `Trainer` will automatcially compute the loss under the torch.cuda.amp.autocast regime, making the forward pass possible to work without any dtype mismatch issue

However, we need to make sure all other methods inside `DPOTrainer` that call the model's forward pass will also use that context manager to make sure there is no dtype mismatch

since this is a regression, will publish a patch release after merging this pr

cc @kashif @pacman100 @lvwerra

### CODE REVIEW

File: trl/trainer/utils.py [hunk 1]
```diff
@@ -641,9 +641,9 @@ def peft_module_casting_to_bf16(model):
     for name, module in model.named_modules():
         if isinstance(module, BaseTunerLayer):
             module = module.to(torch.bfloat16)
-        if "norm" in name:
+        elif isinstance(module, torch.nn.LayerNorm) or "norm" in name:
```
Comment (younesbelkada) [inline trl/trainer/utils.py:644]: @pacman100 is it correct to assume that you want to cast here all layernorms?

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
thanks @younesbelkada awesome!

File: trl/trainer/utils.py (hunk 1 above)
Comment (pacman100) [reply trl/trainer/utils.py:644]: Hello Younes, yes.
